buildSrc/src/main/java/net/neoforged/neodev/GenerateAccessTransformers.java

```diff
@@ -0,0 +1,245 @@
+                current = GenerateAccessTransformers.getClass(targets, name);
+                current.access.setValue(access);
+                if (superName != null) {
+                    current.parents.add(GenerateAccessTransformers.getClass(targets, superName));
+                }
+                for (String iface : interfaces) {
+                    current.parents.add(GenerateAccessTransformers.getClass(targets, iface));
+                }
+            }
+
+            @Override
+            public void visitInnerClass(String name, String outerName, String innerName, int access) {
+                if (name.equals(current.name)) {
+                    current.access.setValue(access);
+                }
+            }
+
+            @Override
+            public FieldVisitor visitField(int access, String name, String descriptor, String signature, Object value) {
+                current.fields.add(new FieldInfo(name, GenerateAccessTransformers.getClass(targets, Type.getType(descriptor).getInternalName()), access));
+                return null;
+            }
+
+            @Override
+            public MethodVisitor visitMethod(int access, String name, String descriptor, String signature, String[] exceptions) {
+                current.addMethod(name, descriptor, access);
+                return null;
+            }
+        };
+    }
+
+    private static ClassInfo getClass(Map<String, ClassInfo> targets, String name) {
+        var existing = targets.get(name);
+        if (existing != null) return existing;
+        existing = new ClassInfo(name, new MutableInt(0), new ArrayList<>(), new ArrayList<>(), new ArrayList<>());
+        targets.put(name, existing);
+        return existing;
+    }
+
+    public record ClassInfo(String name, MutableInt access, List<ClassInfo> parents, List<MethodInfo> methods, List<FieldInfo> fields) {
+        public void addMethod(String name, String desc, int access) {
+            this.methods.add(new MethodInfo(name, desc, access));
+        }
+
+        public boolean hasSuperclass(String name) {
+            for (ClassInfo parent : parents) {
+                if (parent.hasSuperclass(name)) {
+                    return true;
+                }
+            }
+            return this.name.equals(name);
+        }
+    }
+    public record MethodInfo(String name, String descriptor, int access) {}
+    public record FieldInfo(String name, ClassInfo type, int access) {}
+
+    public record AtGroup(String name, Modifier modifier, SerializablePredicate<ClassInfo> classMatch,
+                          @Nullable SerializablePredicate<MethodInfo> methodMatch, @Nullable SerializablePredicate<FieldInfo> fieldMatch) implements Serializable {
+    }
+
+    @FunctionalInterface
+    public interface SerializablePredicate<T> extends Serializable {
+        boolean test(T value);
+    }
+
+    @FunctionalInterface
+    public interface Modifier extends Serializable {
```

Comment:
I would get rid of the interface and instead make it a class directly. We don't expect other instances to be created.